File: src/cloud_recording_manager.py
import os
import logging
from typing import Optional
from datetime import datetime
from livekit import api

logger = logging.getLogger(__name__)

class CloudRecordingManager:
    """
    Manages LiveKit Cloud Recording using Egress API
    Records complete interview (all participants) into single MP4 file
    """
    
    def __init__(self):
        self.egress_id: Optional[str] = None
        self.candidate_id: Optional[str] = None
        
        # Initialize LiveKit API client
        self.livekit_api = api.LiveKitAPI(
            url=os.getenv("LIVEKIT_URL"),
            api_key=os.getenv("LIVEKIT_API_KEY"),
            api_secret=os.getenv("LIVEKIT_API_SECRET")
        )
        
        logger.info("Cloud recording manager initialized")
    
    async def start_recording(self, room_name: str, candidate_id: str):
        """Start cloud recording for the entire room"""
        self.candidate_id = candidate_id
        
        logger.info("=" * 60)
        logger.info(f"STARTING CLOUD RECORDING FOR ROOM: {room_name}")
        logger.info("=" * 60)
        
        try:
            # Create room composite recording (all participants in one video)
            # Inject S3 credentials directly from ENV if available to bypass Console config
            s3_upload = None
            if os.getenv("S3_ACCESS_KEY") and os.getenv("S3_SECRET_KEY") and os.getenv("S3_BUCKET"):
                logger.info("Injecting S3 credentials from environment variables...")
                s3_upload = api.S3Upload(
                    access_key=os.getenv("S3_ACCESS_KEY"),
                    secret=os.getenv("S3_SECRET_KEY"),
                    region=os.getenv("S3_REGION", "us-east-1"),
                    bucket=os.getenv("S3_BUCKET"),
                    endpoint=os.getenv("S3_ENDPOINT") # Optional
                )
            
            file_output = api.EncodedFileOutput(
                file_type=api.EncodedFileType.MP4,
                filepath=f"ai_interview/{candidate_id}/{{egress_id}}.mp4",
                s3=s3_upload
            )
            
            # Start room composite egress (records entire room)
            request = api.RoomCompositeEgressRequest(
                room_name=room_name,
                layout="grid",  # Grid layout shows all participants
                audio_only=False,
                video_only=False,
                file_outputs=[file_output]
            )
            
            egress = await self.livekit_api.egress.start_room_composite_egress(request)
            self.egress_id = egress.egress_id
            
            logger.info(f"✓ Cloud recording started successfully")
            logger.info(f"  Egress ID: {self.egress_id}")
            logger.info(f"  Room: {room_name}")
            logger.info(f"  Recording will be available in LiveKit Cloud Console")
            logger.info("=" * 60)
            
            return egress
            
        except Exception as e:
            if "resource_exhausted" in str(e).lower() or "limit exceeded" in str(e).lower():
                logger.error("!!!" * 20)
                logger.error("LIVEKIT ERROR: CONCURRENT EGRESS LIMIT EXCEEDED")
                logger.error("You have too many active recordings running. Please stop old ones in LiveKit Cloud Consol.")
                logger.error("!!!" * 20)
            else:
                logger.error(f"Failed to start cloud recording: {e}", exc_info=True)
            raise
    
    async def stop_recording(self):
        """Stop the cloud recording"""
        if not self.egress_id:
            logger.warning("No active recording to stop")
            return None
        
        logger.info("=" * 60)
        logger.info("STOPPING CLOUD RECORDING")
        logger.info("=" * 60)
        
        try:
            # Stop the egress
            request = api.StopEgressRequest(egress_id=self.egress_id)
            egress = await self.livekit_api.egress.stop_egress(request)
            
            logger.info(f"✓ Cloud recording stopped successfully")
            logger.info(f"  Egress ID: {self.egress_id}")
            logger.info(f"  Status: {egress.status}")
            
            # Get download info
            if egress.file_results:
                for file_result in egress.file_results:
                    logger.info(f"  Recording file: {file_result.filename}")
                    if file_result.download_url:
                        logger.info(f"  Download URL: {file_result.download_url}")
            
            logger.info("=" * 60)
            logger.info("📥 TO DOWNLOAD YOUR RECORDING:")
            logger.info("1. Go to LiveKit Cloud Console: https://cloud.livekit.io/")
            logger.info("2. Navigate to 'Egress' or 'Recordings' section")
            logger.info(f"3. Find recording with Egress ID: {self.egress_id}")
            logger.info("4. Click 'Download' button")
            logger.info("=" * 60)
            
            return egress
            
        except Exception as e:
            logger.error(f"Error stopping cloud recording: {e}", exc_info=True)
            return None
        finally:
            await self.livekit_api.aclose()
    
    async def get_recording_info(self):
        """Get current recording status and info"""
        if not self.egress_id:
            return None
        
        try:
            egress = await self.livekit_api.egress.list_egress(egress_id=self.egress_id)
            return egress
        except Exception as e:
            logger.error(f"Error getting recording info: {e}")
            return None

File: backend/cloud-recording-manager.js
import { EgressClient, S3Upload, EncodedFileOutput, EncodedFileType } from 'livekit-server-sdk';

export class CloudRecordingManager {
    constructor() {
        this.egressId = null;
        this.candidateId = null;

        this.egressClient = new EgressClient(
            process.env.LIVEKIT_URL,
            process.env.LIVEKIT_API_KEY,
            process.env.LIVEKIT_API_SECRET
        );

        console.log("Cloud recording manager initialized");
    }

    /**
     * Start cloud recording for the entire room.
     *
     * S3 upload notes:
     *  - `endpoint` must be the regional endpoint (e.g. https://s3.ap-south-1.amazonaws.com)
     *    so LiveKit Egress routes to the right AWS region.
     *  - `forcePathStyle: false` is REQUIRED for AWS S3.  When a custom endpoint is set,
     *    LiveKit Egress defaults to path-style (s3.region.amazonaws.com/bucket/key).
     *    AWS returns 301 PermanentRedirect for path-style on new buckets.
     *    Virtual-hosted style (bucket.s3.region.amazonaws.com/key) works correctly.
     */
    async startRecording(roomName, candidateId, filename) {
        this.candidateId = candidateId;

        console.log("=".repeat(60));
        console.log(`STARTING CLOUD RECORDING FOR ROOM: ${roomName}`);
        console.log("=".repeat(60));

        try {
            let s3Upload = null;
            if (process.env.S3_ACCESS_KEY && process.env.S3_SECRET_KEY && process.env.S3_BUCKET) {
                const region = process.env.S3_REGION || "us-east-1";
                s3Upload = new S3Upload({
                    accessKey: process.env.S3_ACCESS_KEY,
                    secret: process.env.S3_SECRET_KEY,
                    region,
                    endpoint: `https://s3.${region}.amazonaws.com`,
                    bucket: process.env.S3_BUCKET,
                    forcePathStyle: false,   // use virtual-hosted style: bucket.s3.region.amazonaws.com
                });
            }

            const basePath = `ai_interview/${candidateId}/${filename}.mp4`;

            const fileOutput = new EncodedFileOutput({
                fileType: EncodedFileType.MP4,
                filepath: basePath,
                ...(s3Upload && { output: { case: 's3', value: s3Upload } })
            });

            const egress = await this.egressClient.startRoomCompositeEgress(
                roomName,
                fileOutput,
                { layout: "grid" }
            );

            this.egressId = egress.egressId;

            console.log(`✓ Cloud recording started successfully`);
            console.log(`  Egress ID: ${this.egressId}`);
            console.log(`  Room: ${roomName}`);
            console.log(`  S3 path: s3://${process.env.S3_BUCKET}/${basePath}`);
            console.log("=".repeat(60));

            return egress;

        } catch (error) {
            console.error(`Failed to start cloud recording:`, error);
            throw error;
        }
    }

    /**
     * Stop the cloud recording
     */
    async stopRecording() {
        if (!this.egressId) {
            console.warn("No active recording to stop");
            return null;
        }

        console.log("=".repeat(60));
        console.log("STOPPING CLOUD RECORDING");
        console.log("=".repeat(60));

        try {
            const egress = await this.egressClient.stopEgress(this.egressId);

            console.log(`✓ Cloud recording stopped successfully`);
            console.log(`  Egress ID: ${this.egressId}`);
            console.log(`  Status: ${egress.status}`);

            console.log("=".repeat(60));
            console.log("📥 TO VERIFY YOUR RECORDING:");
            console.log("1. Go to LiveKit Cloud Console: https://cloud.livekit.io/");
            console.log(`2. Find recording with Egress ID: ${this.egressId}`);
            console.log(`3. Or check S3 bucket: s3://${process.env.S3_BUCKET}/ai_interview/${this.candidateId}/`);
            console.log("=".repeat(60));

            return egress;

        } catch (error) {
            console.error(`Error stopping cloud recording:`, error);
            return null;
        }
    }
}
